### ISSUE TITLE
Calculate_logical_size panic with `should had found the parent`

### ISSUE THREAD

**Issue by lubennikovaav** [neondatabase/neon#3179]
TLDR: Calculate_logical_size panic with `should had found the parent` because it expects ordered list of branches, where it isn't.

## Steps to reproduce

run `test_metric_collection_multiple_branches` from https://github.com/neondatabase/neon/pull/3123
On my laptop error reproduces every second run or so.

## Expected result
No panic

## Actual result

calculate_logical_size panics with `should had found the parent`

I've added some debug messages in my PR and moved error handling. See log below.

This happens because `branch` command cannot find a parent.
But the updates vec is not sorted in any particular order and doesn't guarantee that we won't try to insert child before parent.

## Logs, links

```
2022-12-21T22:56:37.730127Z  INFO metrics_collection:calculate_synthetic_size{tenant_id=e206d18c482b42f94e98ff9e4d0e9926}: insert point for branch Some(cfbd83f2bfb27b3fc5f1b8dbc42ff8d4) newseg_id 1    
2022-12-21T22:56:37.730153Z  INFO metrics_collection:calculate_synthetic_size{tenant_id=e206d18c482b42f94e98ff9e4d0e9926}: insert point for branch Some(cfbd83f2bfb27b3fc5f1b8dbc42ff8d4) newseg_id 1    
thread 'mgmt request worker' panicked at 'err should had found the parent Some(cde8e669415fdb999018d456ad6ff94a) by key. in branches {None: 0, Some(cfbd83f2bfb27b3fc5f1b8dbc42ff8d4): 1}, inputs ModelInputs { updates: [Update { lsn: 0/0, command: BranchFrom(None), timeline_id: cfbd83f2bfb27b3fc5f1b8dbc42ff8d4 }, Update { lsn: 0/169AD58, command: Update(25387008), timeline_id: cfbd83f2bfb27b3fc5f1b8dbc42ff8d4 }, Update { lsn: 0/169AD58, command: BranchFrom(Some(cde8e669415fdb999018d456ad6ff94a)), timeline_id: 095ffdc034f328fd3ec9a012a2b8bc8d }, Update { lsn: 0/169AD58, command: BranchFrom(Some(cfbd83f2bfb27b3fc5f1b8dbc42ff8d4)), timeline_id: cde8e669415fdb999018d456ad6ff94a }, Update { lsn: 0/1CA85B8, command: Update(28925952), timeline_id: 095ffdc034f328fd3ec9a012a2b8bc8d }, Update { lsn: 0/1CD85B8, command: Update(29024256), timeline_id: 095ffdc034f328fd3ec9a012a2b8bc8d }, Update { lsn: 0/1CD85B8, command: BranchFrom(Some(095ffdc034f328fd3ec9a012a2b8bc8d)), timeline_id: 26dd492d88b5f00deaca94faf87e64a0 }, Update { lsn: 0/22DCE70, command: Update(32546816), timeline_id: 26dd492d88b5f00deaca94faf87e64a0 }, Update { lsn: 0/230CE70, command: EndOfBranch, timeline_id: 26dd492d88b5f00deaca94faf87e64a0 }], retention_period: 196608, timeline_inputs: {095ffdc034f328fd3ec9a012a2b8bc8d: TimelineInputs { last_record: 0/1CD85B8, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/1CA85B8, pitr_cutoff: 0/1CA85B8, next_gc_cutoff: 0/1CA85B8 }, cde8e669415fdb999018d456ad6ff94a: TimelineInputs { last_record: 0/169AD58, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/166AD58, pitr_cutoff: 0/166AD58, next_gc_cutoff: 0/166AD58 }, 26dd492d88b5f00deaca94faf87e64a0: TimelineInputs { last_record: 0/230CE70, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/22DCE70, pitr_cutoff: 0/22DCE70, next_gc_cutoff: 0/22DCE70 }, cfbd83f2bfb27b3fc5f1b8dbc42ff8d4: TimelineInputs { last_record: 0/169AD58, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/166AD58, pitr_cutoff: 0/166AD58, next_gc_cutoff: 0/166AD58 }} }'
```

@koivunej , FYI

**Comment by lubennikovaav** [comment]
I don't know how to sort branches, given that the graph can be complex.
So I think we can just add them all into `storage` without checks, and then do another iteration on `storage` to ensure that all entries are present and linked correctly.

**Comment by SomeoneToIgnore** [comment]
https://github.com/neondatabase/neon/blob/7ff591ffbfc8f084e0b1b5cdbed8bd69e008d4c0/pageserver/src/tenant.rs#L1527-L1530

Could that help?

**Comment by koivunej** [comment]
Yeah the version I wrote doesn't by design use tree_sort_timelines. It was as simple as possible to allow further experimentation, latest of which is now #2817.

As quickly chatted about this, I don't have time to look into this right now, but I'd understand the sorted output. If that means that we are witnessing a timeline being-created then that should probably just be filtered away. Of course, there might be other issues caused by updating gc info on such timeline.

I'd guess that timeline being-created has `latest_lsn == 0` which is less than it's `ancestor_lsn`, so this might work as a filter.

**Comment by koivunej** [comment]
Okey, unpacking the panic message looks like:

```
should had found the parent Some(cde8e669415fdb999018d456ad6ff94a) by key {None: 0, Some(cfbd83f2bfb27b3fc5f1b8dbc42ff8d4): 1}, inputs 

ModelInputs {
  updates: [
    Update { lsn: 0/0, command: BranchFrom(None), timeline_id: cfbd83f2bfb27b3fc5f1b8dbc42ff8d4 },
    Update { lsn: 0/169AD58, command: Update(25387008), timeline_id: cfbd83f2bfb27b3fc5f1b8dbc42ff8d4 },
    # if these two lines were sorted differently this would work
    Update { lsn: 0/169AD58, command: BranchFrom(Some(cde8e669415fdb999018d456ad6ff94a)), timeline_id: 095ffdc034f328fd3ec9a012a2b8bc8d },
    Update { lsn: 0/169AD58, command: BranchFrom(Some(cfbd83f2bfb27b3fc5f1b8dbc42ff8d4)), timeline_id: cde8e669415fdb999018d456ad6ff94a },
    Update { lsn: 0/1CA85B8, command: Update(28925952), timeline_id: 095ffdc034f328fd3ec9a012a2b8bc8d },
    Update { lsn: 0/1CD85B8, command: Update(29024256), timeline_id: 095ffdc034f328fd3ec9a012a2b8bc8d },
    Update { lsn: 0/1CD85B8, command: BranchFrom(Some(095ffdc034f328fd3ec9a012a2b8bc8d)), timeline_id: 26dd492d88b5f00deaca94faf87e64a0 },
    Update { lsn: 0/22DCE70, command: Update(32546816), timeline_id: 26dd492d88b5f00deaca94faf87e64a0 },
    Update { lsn: 0/230CE70, command: EndOfBranch, timeline_id: 26dd492d88b5f00deaca94faf87e64a0 }
  ],
  retention_period: 196608,
  timeline_inputs: {
    095ffdc034f328fd3ec9a012a2b8bc8d: TimelineInputs { last_record: 0/1CD85B8, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/1CA85B8, pitr_cutoff: 0/1CA85B8, next_gc_cutoff: 0/1CA85B8 },
    cde8e669415fdb999018d456ad6ff94a: TimelineInputs { last_record: 0/169AD58, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/166AD58, pitr_cutoff: 0/166AD58, next_gc_cutoff: 0/166AD58 },
    26dd492d88b5f00deaca94faf87e64a0: TimelineInputs { last_record: 0/230CE70, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/22DCE70, pitr_cutoff: 0/22DCE70, next_gc_cutoff: 0/22DCE70 },
    cfbd83f2bfb27b3fc5f1b8dbc42ff8d4: TimelineInputs { last_record: 0/169AD58, latest_gc_cutoff: 0/169ACE0, horizon_cutoff: 0/166AD58, pitr_cutoff: 0/166AD58, next_gc_cutoff: 0/166AD58 }
  } 
}
```

Yeah, this is not a timeline being-created issue I was thinking earlier, it's a "missing tree_sort" issue. I did not know this kind of situation could happen.
